python/src/iceberg/avro/resolver.py

```diff
@@ -0,0 +1,150 @@
+    NoneReader,
+    OptionReader,
+    Reader,
+    StructReader,
+    primitive_reader,
+)
+from iceberg.schema import Schema, visit
+from iceberg.types import (
+    BinaryType,
+    IcebergType,
+    IntegerType,
+    ListType,
+    LongType,
+    MapType,
+    PrimitiveType,
+    StringType,
+    StructType,
+)
+
+
+class ResolveException(Exception):
+    pass
+
+
+@singledispatch
+def resolve(write_schema, read_schema) -> Reader:
+    """This resolves the write and read schema
+
+    The function traverses the schema in post-order fashion
+
+     Args:
+         write_schema (Schema | IcebergType): The write schema of the Avro file
+         read_schema (Schema | IcebergType): The requested read schema which is equal or a subset of the write schema
+
+     Raises:
+         NotImplementedError: If attempting to resolve an unrecognized object type
+    """
+    raise NotImplementedError("Cannot resolve non-type: %s" % write_schema)
+
+
+@resolve.register(Schema)
+def _(write_schema: Schema, read_schema: Schema) -> Reader:
+    """Visit a Schema and starts resolving it by converting it to a struct"""
+    return resolve(write_schema.as_struct(), read_schema.as_struct())
+
+
+@resolve.register(StructType)
+def _(write_struct: StructType, read_struct: StructType) -> Reader:
+    """Iterates over the write schema, and checks if the field is in the read schema"""
+    results: List[Tuple[Optional[int], Reader]] = []
+
+    read_fields = {field.name: (pos, field) for pos, field in enumerate(read_struct.fields)}
+
+    for write_field in write_struct.fields:
+        if write_field.name in read_fields:
+            read_pos, read_field = read_fields[write_field.name]
+            result_reader = resolve(write_field.field_type, read_field.field_type)
+        else:
+            read_pos = None
+            result_reader = visit(write_field.field_type, ConstructReader())
+        result_reader = result_reader if write_field.required else OptionReader(result_reader)
+        results.append((read_pos, result_reader))
+
+    write_fields = {field.name: field for field in write_struct.fields}
+    for pos, read_field in enumerate(read_struct.fields):
+        if read_field.name not in write_fields:
+            if read_field.required:
+                raise ResolveException(f"{read_field} is in not in the write schema, and is required")
+            # Just set the new field to None
+            results.append((pos, NoneReader()))
+
+    return StructReader(tuple(results))
+
+
+@resolve.register(ListType)
+def _(write_list: ListType, read_list: ListType) -> Reader:
+    if not isinstance(read_list, ListType):
+        raise ResolveException(f"Cannot change {write_list} into {read_list}")
+    element_reader = resolve(write_list.element.field_type, read_list.element.field_type)
+    return ListReader(element_reader)
+
+
+@resolve.register(MapType)
+def _(write_map: MapType, read_map: MapType) -> Reader:
+    if not isinstance(read_map, MapType):
```

Comment:
`read_map` is declared to be a `MapType`, but since you're checking it here it seems like it should be `IcebergType` instead? That would match what is done for `PrimitiveType`, although I'm not sure how you get type checking to recognize that it's a `MapType` after all.

Reply:
Yes, I think `IcebergType` would be better as the second argument, let me update that. The first one is guaranteed because that's what singledispatch gives us. We could have a write schema:
```
read schema {
  100: location struct<101: lat double, 102: long double>
}
```
and a read schema:
```
read schema {
  100: long
}
```
I wanted to have a nicer error in this case than just an error that `key` can't be found (because a `LongType` doesn't have a key property).

Reply:
I've updated this for all the complex types 👍🏻

Reply:
Yeah, I agree with having a nice error message if the types don't match!